### ISSUE TITLE
Rule request: forbid control flow statements in finally

### ISSUE THREAD

**Issue by louy** [eslint/eslint#5808]
Adding a `return` statement in a `finally` block is a very confusing thing. Thrown errors get ignored, `return` statements in `try` blocks get ignored, and your mind explodes. 

I'd love to have a rule to forbid this. Especially now that we're having `async`/`await`.

examples:
```js
try {
  return 1;
} catch(err) {
  return 2;
} finally {
  return 3;
}
// > 3

try {
  throw new Error;
  return 1;
} catch(err) {
  return 2;
} finally {
  return 3;
}
// > 3
```

**Comment by mysticatea** [comment]
Thank you for this issue.

Indeed, ReturnStatements in a `finally` block may hide actual errors or results.
I think also ThrowStatements is similar to it.

I like this idea, but we need to wait for opinions of other members.

**Comment by BYK** [comment]
I'm :+1: since this is a core language issue and a pitfall quite easy to fall into without realizing consequences. We should add this to `eslint:recommended` too IMO.

**Comment by nzakas** [comment]
Seems like a good idea. Does anyone want to champion this? @eslint/eslint-team 

@byk we only change `eslint:recommended ` when we do a major release because it's a breaking change.

**Comment by michaelficarra** [comment]
If `return` is forbidden, all control flow operations should be forbidden: `throw`, `break`, and `continue`.

**Comment by platinumazure** [comment]
:+1: to @michaelficarra. Perhaps the rule could be called something like `no-finally-control-flow`?

**Comment by BYK** [comment]
@nzakas - I know. I wanted to raise the flag to include this in 3.0.0 planning :)

I can champion the rule. @onurtemizkan interested in implementing?

**Comment by onurtemizkan** [comment]
Sure! :smile:

**Comment by nzakas** [comment]
Okay @byk is championing this rule. Please work with @onurtemizkan to implement the rule using the [new format](https://github.com/eslint/eslint/blob/master/docs/developer-guide/working-with-rules-new.md).

**Comment by BYK** [comment]
Did we decide on a rule name?

`no-finally-control-flow` sounds like a good one but a bit on the longer side. Other suggestions: `no-control-finally`, `no-control-in-finally`, `no-unpredictable-finally`.

@onurtemizkan has started working on this.

**Comment by platinumazure** [comment]
I could probably get behind any of those. I'll throw one more name into the
ring: `no-interrupt-finally`.

**Comment by nzakas** [comment]
"Control flow" refers to anything that changes where execution flows, Including `if`, `while`, etc., so it's a bit broad for this use case. 

How about `no-unsafe-finally`?

**Comment by BYK** [comment]
`no-unsafe-finally` may be a bit too broad but I like it.

**Comment by nzakas** [comment]
Okay, we'll leave this to @byk and @onurtemizkan to implement.

**Comment by onurtemizkan** [comment]
@michaelficarra I couldn't find a case for `break` and `continue`, creating unexpected behaviour in `finally`. Could you give an example?

**Comment by michaelficarra** [comment]
```js
(function() {
  label: try {
    return 0;
  } finally {
    break label;
  }
  return 1;
})()
```

**Comment by onurtemizkan** [comment]
Ah, okay. Thanks!

**Comment by bakkot** [comment]
Currently this rule isn't actually checking that breaks and continues are crossing the boundary of the `finally`. In fact, one of the examples of correct code from the docs actually gives an errors:

```js
let foo = function(a) {
    try {
        return 1;
    } catch(err) {
        return 2;
    } finally {
        switch(a) {
            case 1: {
                console.log("hola!")
                break;
            }
        }
    }
};
```

This also trips this rule, I think incorrectly:
```js
try {} finally {
  while(true) {
    break;
  }
}
```

I would not think the mere presence of control flow statements would be an issue: it's only control flow which crosses the boundary of the `finally`. Doing this right involves determining which statements are broken / continued.

I can open an issue for this if people agree.

**Comment by platinumazure** [comment]
@bakkot I agree with your logic. Please open a new issue and I will triage it as a bug right away. :smile:

**Comment by onurtemizkan** [comment]
@bakkot Sorry, my bad. I can post a fix for this if you open an issue.

**Comment by bakkot** [comment]
Opened an issue.

**Comment by mk-pmb** [comment]
Does anyone have an in-the-wild example where flow control in a finally block caused an actual problem?
